Fix tests/application_development/code_relocation for mr_canhubk3

manuargue opened the pull request:
Add a memory region for the IVT header on NXP S32K3 devices. This is normally done from the SoC linker script but this test make use of a custom linker script.

In some platforms the MPU region used for NULL pointer detection conflicts with the ITCM region, causing access fault. Make disabling the null pointer exception detection the default behavior for the Arm ITCM relocation test.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/60165

manuargue commented:
I've added a commit to skip this test for `nucleo_h7a3zi_q`, because is failing to build due to unrelated reasons, even before these changes. Logged it to #60167

hakehuang commented on tests/application_development/code_relocation/linker_arm_sram2.ld:
```diff
@@ -34,6 +34,11 @@ MEMORY
+#ifdef CONFIG_SOC_SERIES_S32K3_M7
+    IVT_HEADER              (r) : ORIGIN = (CONFIG_FLASH_BASE_ADDRESS + CONFIG_IVT_HEADER_OFFSET),
+                                  LENGTH = CONFIG_IVT_HEADER_SIZE
+#endif
```
better to move the changes to board config, directly to another linker script for S32 series, if they are not using
#include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>

manuargue replied:
This SoC is using the arch linker file https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_s32/s32k/linker.ld#L15. But this test is using a custom linker script, therefore not using the ones at SoC level.

hakehuang replied:
@manuargue , for such platform specific case, you need create a new linker script and add a new config and change the testcase.yml create a new case for S30 platform. the linker_arm_sram2.ld is based on platform that using the #include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>

manuargue replied:
S32K3 SoC is using the arch linker `#include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>` already, as I pointed out in https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_s32/s32k/linker.ld#L15. The issue is this test is using  `CONFIG_CUSTOM_LINKER_SCRIPT="linker_arm_sram2.ld"` which overrides the SoC linker so the needed memory region is not present, and that's why is introduced here. I could move it to a separate linker file, but I consider is still part of Arm test, and not a special case. There are many platforms that define a SoC specific linker, so I wonder on which SoCs this test is actually running.

erwango commented on tests/application_development/code_relocation/testcase.yaml:
```diff
@@ -6,6 +6,9 @@ tests:
     extra_configs:
       - CONFIG_RELOCATE_TO_ITCM=y
+      - CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
+    platform_exclude:
+      - nucleo_h7a3zi_q
```
I rather think that the issue is in the `filter/arch_allow` combination.
Looking to `tests/application_development/code_relocation/CMakeLists.txt` I wonder how the following twister filter is supposed to produce meaningful results on platforms other than NXP.
`zephyr,itcm` is not available only on NXP platforms (nucleo_h7a3zi_q is a good example), yet test will be run on any platform supporting `zephyr,itcm`.
`platform_exclude` is a one bullet gun. It will have to be updated each time wrongly composed `filter/arch_allow` fails to provide correct filtering

hakehuang replied:
I port this testcases due to an issue found when code relocate in pre-kernel memcpy usage, so it is limited to NXP platforms, but following enhances seems try to introduce more platforms. But I think this is not related to this PR

erwango replied:
Problem is I don't understand how this is supposed to work on platforms that are not dealt with in CMakeLists.txt. IMO filter should be much more restrictive.

manuargue replied:
I agree that filters should be more restrictive if this test is meant only to some NXP platforms, but I'd suggest @hakehuang (as initial author of this test) address this in a separate PR to address #60167 and for the many other SoCs this test may not be building. If there's a limited subset of NXP boards this test is addressed, then a `allow_platforms` filter seems more reasonable here.

manuargue replied:
> platform_exclude is a one bullet gun. It will have to be updated each time wrongly composed filter/arch_allow fails to provide correct filtering

Right, the addition of `platform_exclude` here is not meant to be a fix, just a workaround to unblock 6 other PR's affected by this test failure. I'm not familiar with the ST boards so I cannot propose a fix for them and logged it separately in #60167. As I commented there, this is likely to happen in many other Arm boards as well.

erwango replied:
I think the solution is to restrict to NXP boards for now

hakehuang replied:
then we need add platform_allow

manuargue replied:
@hakehuang `frdm_k64f` is not selecting any `zephyr,itcm`. This test cannot link for that board. Can you please double check this is correct?

hakehuang replied:
@manuargue , Thanks a lot for this checking, it was a mistaken here please do below changes, and I test is on frdm_k64f whcih works
```
  tests.application_development.code_relocation_kinetis:
    filter: CONFIG_CPU_HAS_NXP_MPU and CONFIG_MINIMAL_LIBC
    arch_allow: arm
    extra_configs:
      - CONFIG_MPU_ALLOW_FLASH_WRITE=y
    platform_allow:
      - frdm_k64f
```

```
*** Booting Zephyr OS build zephyr-v3.4.0-975-gd8b2a7263ba8 ***
Running TESTSUITE code_relocation
===================================================================
START - test_function_genex_relocate_1
Address of var_file4_sram2_data 0x200193a4
Address of var_file4_sram2_bss 0x200193e0

 PASS - test_function_genex_relocate_1 in 0.008 seconds
===================================================================
START - test_function_genex_relocate_2
Address of var_file5_sram2_data 0x200193a8
Address of var_file5_sram2_bss 0x200193f8

 PASS - test_function_genex_relocate_2 in 0.008 seconds
===================================================================
START - test_function_in_split_multiple
Address of var_file3_sram_data 0x20000580
Address of var_file3_sram2_bss 0x200193c8

 PASS - test_function_in_split_multiple in 0.008 seconds
===================================================================
START - test_function_in_sram2
Address of var_sram2_data 0x200193a0
Address of k_sem_give 0x200181c1
Address of var_sram2_rodata 0x20018b40
Address of var_sram2_bss 0x200193b0

Hello World! frdm_k64f
Address of memcpy 0x6303

Address of var_lib1_sram2_data 0x200193ac
Address of var_lib1_sram2_bss 0x20019410
Address of relocated_lib_helper 0x200185fd

Relocated helper function running on frdm_k64f

Address of custom_section, func placed using attributes 0x20001b85
Address of custom_section data placed using attributes 0x20001b80

 PASS - test_function_in_sram2 in 0.046 seconds
===================================================================
TESTSUITE code_relocation succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [code_relocation]: pass = 4, fail = 0, skip = 0, total = 4 duration = 0.070 seconds
 - PASS - [code_relocation.test_function_genex_relocate_1] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_genex_relocate_2] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_in_split_multiple] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_in_sram2] duration = 0.046 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

```

manuargue-nxp committed:
tests: code_relocation: restrict Arm test to NXP platforms

Currently tests.application_development.code_relocation and
tests.application_development.code_relocation_kinetis scenarios are
restricted to specific NXP platforms only, so make filters more strict.

Also for frdm_k64f, which lacks of ITCM, relocation to ITCM must be
disabled.

Fixes #60167

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
tests: code_relocation: support mr_canhubk3 boards

Add test scenario for NXP S32 platforms, for now only ported to
mr_canhubk3.

In some platforms the MPU region used for NULL pointer detection
conflicts with the ITCM region, causing access fault. Make disabling the
null pointer exception detection the default behavior for this scenario.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
@erwango with this should we close also #60167? or do you want to keep it opened to support this test later on ST boards?

erwango commented:
> @erwango with this should we close also #60167? or do you want to keep it opened to support this test later on ST boards?

We can close it. If an issue is needed, this should be a feature request, not a bug.

manuargue commented:
> > @erwango with this should we close also #60167? or do you want to keep it opened to support this test later on ST boards?
> 
> We can close it. If an issue is needed, this should be a feature request, not a bug.

ok, re-pushed to add `fixes #` to each commit. Thanks

hakehuang commented on tests/application_development/code_relocation/linker_arm_nxp_s32.ld:
```diff
@@ -0,0 +1,31 @@
+#endif
+    }
+
+#include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>
```
should be the s32 specific linker file?

manuargue replied:
this is what I mentioned before in https://github.com/zephyrproject-rtos/zephyr/pull/60166#discussion_r1258394185.
The linker file for the Arm test is made in such a way that it doesn't scale if we want to support more SoCs that use SoC-specific linker files (which are most of them). The reason it works for the two boards you ported it, frdm_k64f and mimxrt1060_evk, is because the linker files of those SoCs are just including `#include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>` (https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_kinetis/k6x/linker.ld, https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_imx/rt/linker.ld). There's nothing special on the NXP S32 platforms with respect of the other Arm SoCs that warrants to use a different linker file as requested.

I think the solution here is to replace `#include <zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld>` with `#include <linker.ld>` in `linker_arm_sram2.ld`, to extend the SoC-specific linker file and not the arch default linker file.

hakehuang replied:
I think this is a good change

manuargue-nxp committed:
tests: code_relocation: use per-SoC linker file for Arm tests

In order to support more Arm SoCs, extend the SoC-specific linker file
instead of the default arch linker file.

Fixes #60165

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
@hakehuang all comments addressed, please help me with a final review. Thanks
